[DynamicCast] Rely on runtime when casts can't be optimized

tbkka opened the pull request:
The Swift compiler renders `as?` operations in Swift into variations of the `checked_cast_br` instructions in SIL.  Subsequent optimization passes may alter or eliminate these instructions.  Any remaining cast instructions after optimization are translated by IRGen into runtime calls.

At least, that's the theory.  Unfortunately, the current IRGen logic does not recognize all of the casting instruction variants and translates one particular unrecognized variant into a simple nil load.  This effectively converts an `as?` operation into one that will always fail.  Specifically, this occurs for optimized casts of metatypes to AnyObject:
```
  let a = Int.self
  let b = a as? AnyObject
  // b should not be nil here
```
(Ironically, in this particular case, the frontend issues a diagnostic warning that the cast will always succeed.)

This PR changes this case in IRGen to instead generate a call to `swift_dynamicCast`, deferring this case to the runtime.  (Yes, this particular example should get optimized into a simple store, but that's a future project.)

Test suite changes:
* With this change, the "BoxingCasts" test suite (which exercises a variety of boxing and unboxing casts) now passes in both optimized and non-optimized builds, so I've enabled those versions.
* There was a SIL test previously that constructed a `checked_cast_br` instruction with a thin metatype.  The IRGen pass should never see instructions of this form, so I've changed that particular test to exercise a thick metatype instead.

tbkka committed:
[DynamicCast] Rely on runtime when casts can't be optimized

The Swift compiler renders `as?` operations in Swift into variations of the
`checked_cast_br` instructions in SIL.  Subsequent optimization passes may alter
or eliminate these instructions.  Any remaining instructions after optimization
are translated by IRGen into runtime calls.

At least, that's the theory.  Unfortunately, the current IRGen logic does not
recognize all of the casting instruction variants and renders one particular
unrecognized variant as a simple nil load.  Specifically, this occurs for
optimized casts of metatypes to AnyObject:
```
  let a = Int.self
  let b = a as? AnyObject
  // b should not be nil here
```

This PR changes this case in IRGen to instead generate a call to
`swift_dynamicCast`, deferring this case to the runtime.

Future:  Someday, the compiler should be taught to correctly optimize
this cast away.

tbkka commented:
@swift-ci Please test

jckarter reviewed (approved):
Makes sense. We shouldn't try to peephole these in IRGen; we should let the cast optimizer do that if possible.

tbkka commented:
@swift-ci Please test macOS

tbkka commented:
@swift-ci Please test

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

tbkka commented:
Looks like the Linux test is failing in optimized builds due to some issue with ARC.  I'm trying to isolate whether it's an issue with my change or is an existing bug in the compiler ARC optimizations.

tbkka commented:
Thanks to @eeckstein for fixing the ARC bug I was tripping over.   Now let's try those tests again!

tbkka commented:
@swift-ci Please test

tbkka commented:
Linux is crashing on the optimized form of this test:
```
[ RUN      ] BoxingCasts.test184(): Casting SwiftValueBox(AnyHashable(ClassInt)) to ClassInt
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] BoxingCasts.test184(): Casting SwiftValueBox(AnyHashable(ClassInt)) to ClassInt
```

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
Re-ran the full Linux test locally and it's passing for me.  Hmmm....  I'll try the CI bot again and see if that still fails before I spend time digging further.

tbkka commented:
Re-ran the Linux test locally on Ubuntu 16.04 (required upgrading clang-3.8 to clang-12), and it all passes just fine.  I guess I'll try again with the CI preset and see??

tbkka commented:
Still having trouble reproducing this locally.  I'll try again with the exact build options used on the CI bot.

tbkka commented:
@swift-ci Please test

tbkka committed:
Enable another casting test

I think this exercises the same optimizer path that's breaking
another test I've been staring at.

tbkka commented:
@swift-ci Please test

tbkka commented:
Let's see if the test that's been failing on Linux also fails on macOS when you enable it properly.

tbkka commented:
Same cast was in a disabled test in Casts.swift, so I also enabled that unconditionally.

With that, I can now consistently reproduce it on Linux.  The root cause seems to be another variation of the ARC issue that @eeckstein fixed for me earlier.  As before, it fails on Linux because it triggers `__SwiftValue` boxing which behaves slightly differently on Linux than on macOS.

tbkka commented:
I believe #34383 will resolve the last test failure here.

tbkka commented:
apple/swift#34383
@swift-ci Please test

tbkka commented:
Now that #34383 has landed, let's try this again.

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please clean test Linux
